tools/Voron.Recovery/RecoveredDatabaseCreator.cs

```diff
@@ -0,0 +1,427 @@
+        {
+            _logger = logger;
+            _recoveryLogCollection = $"RecoveryLog-{recoverySession}";
+            _orphanAttachmentsDocIdPrefix = $"OrphanAttachments/{recoverySession}";
+            _orphanCountersDocIdPrefix = $"OrphanCounters/{recoverySession}";
+            _logDocId = $"Log/{recoverySession}";
+            _database = documentDatabase;
+            _databaseDestination = new DatabaseDestination(documentDatabase);
+            _results = new SmugglerResult();
+            _databaseDestination.Initialize(new DatabaseSmugglerOptionsServerSide(), _results, ServerVersion.Build);
+            _contextDisposal = _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+            _byteStringContext = new ByteStringContext(SharedMultipleUseFlag.None);
+            Slice.From(_byteStringContext, "Attachments", ByteStringType.Immutable, out _attachmentsSlice);
+
+            using (var tx = _context.OpenWriteTransaction())
+            {
+                using (var doc = _context.ReadObject(
+                    new DynamicJsonValue
+                    {
+                        ["RecoverySession"] = recoverySession,
+                        ["RecoveryStarted"] = DateTime.Now,
+                        [Raven.Client.Constants.Documents.Metadata.Key] = new DynamicJsonValue {[Raven.Client.Constants.Documents.Metadata.Collection] = _recoveryLogCollection}
+                    }, _logDocId, BlittableJsonDocumentBuilder.UsageMode.ToDisk)
+                )
+                {
+                    _database.DocumentsStorage.Put(_context, _logDocId, null, doc);
+                }
+                tx.Commit();
+            }
+        }
+
+        public void WriteDocument(Document document)
+        {
+            var actions = _databaseDestination.Documents();
+            WriteDocumentInternal(document, actions);
+        }
+
+        public void WriteRevision(Document document)
+        {
+            if (_database.DocumentsStorage.RevisionsStorage.Configuration == null)
+            {
+                var revisionsConfiguration = new RevisionsConfiguration {Default = new RevisionsCollectionConfiguration {Disabled = false}};
+                var configurationJson = EntityToBlittable.ConvertCommandToBlittable(revisionsConfiguration, _context);
+                (long index, _) = _database.ServerStore.ModifyDatabaseRevisions(_context, _database.Name, configurationJson, Guid.NewGuid().ToString()).Result;
+                AsyncHelpers.RunSync(() => _database.RachisLogIndexNotifications.WaitForIndexNotification(index, _database.ServerStore.Engine.OperationTimeout));
+            }
+            var actions = _databaseDestination.RevisionDocuments();
+            WriteDocumentInternal(document, actions);
+        }
+
+        public void WriteCounterItem(CounterGroupDetail counterGroupDetail, bool skipDocExistence = false)
+        {
```

Comment:
Let's create a separate class to handle each one of those. At this moment it is hard to follow and maintain.

Reply:
I am not following... WriteCounterItem, WriteDocument, WriteRevisions, WriteConflicts and WriteAttachments exists.
They are 99% of this class any how? (or do you mean to split something else to a different class?)

Reply:
Counter actions wrap in CountersRecover, Document actions wrap in DocumentRecovery, etc. A bit of OOP haven't hurt anyone.

Reply:
Done here : https://github.com/ravendb/ravendb/pull/10266/commits/15db3036b2bbecff6f08aa46fdc062efb8ae55f6

I am not so sure it is needed though.. In my eyes it brings more reading complexity.. 
And there is no real _pattern_ here.  (unless we want to revise Recovery.cs which I guess we do not)